### PR TITLE
Go to reference when there's only one

### LINKED ISSUE #4796: Go to reference when there's only one
### Check for existing issues

- [X] Completed

### Describe the feature

When there's only one reference in "Find all references", go to that reference directly instead of opening a reference buffer.

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by wasd96040501**
close #4796
Similar to implementation in `navigate_to_definitions`
Because find all references  in rust-analyzer will return locations containing location we passed in, so we need remove it.
And I'm not sure is it the correct way to do this(line 7299~7307)

### CODE REVIEW

**Comment by wasd96040501** [comment]
@cla-bot check

File: rust-toolchain.toml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [toolchain]
-channel = "1.75"
```
Comment (osiewicz) [inline rust-toolchain.toml:2]: What's up with that change?

Comment (wasd96040501) [reply rust-toolchain.toml:2]: I cannot compile this repo without changing 1.75 to 1.75.0 and I have no idea about it. I will change it back to 1.75 now.

Comment (osiewicz) [reply rust-toolchain.toml]: Would you mind opening an issue with the logs included? I'll take a look at it.

Comment (wasd96040501) [reply rust-toolchain.toml]: Sure

Comment (wasd96040501) [reply rust-toolchain.toml]: @osiewicz Issue opened: https://github.com/zed-industries/zed/issues/6925

File: crates/editor/src/editor.rs [hunk 2]
```diff
@@ -7287,17 +7287,58 @@ impl Editor {
...
         let workspace = self.workspace()?;
         let project = workspace.read(cx).project().clone();
         let references = project.update(cx, |project, cx| project.references(&buffer, head, cx));
-        Some(cx.spawn(|_, mut cx| async move {
-            let locations = references.await?;
+        Some(cx.spawn(|editor, mut cx| async move {
+            let mut locations = references.await?;
+
+            // Lsp may return references that contain self pos (eg. rust-analyzer)
+            // So we remove self pos from locations
```
Comment (mrnugget) [inline crates/editor/src/editor.rs:7300]: Can you make the comment a little bit clearer? What's a `self pos`, for example? The definition itself?

File: crates/editor/src/editor.rs [hunk 3]
```diff
@@ -7287,17 +7287,58 @@ impl Editor {
...
+            editor.update(&mut cx, |_, cx| {
```
Comment (mrnugget) [inline crates/editor/src/editor.rs:7301]: Does this need to be in an `editor.update`?

File: crates/editor/src/editor.rs [hunk 4]
```diff
@@ -7287,17 +7287,58 @@ impl Editor {
...
+                                });
+                            target_editor.update(cx, |target_editor, cx| {
+                                target_editor.change_selections(Some(Autoscroll::fit()), cx, |s| {
+                                    s.select_ranges([range]);
+                                })
+                            })
+                        })
+                    }
+                })?;
+                return Ok(());
```
Comment (mrnugget) [inline crates/editor/src/editor.rs:7339]: You can return the value of `editor.update`, right?

File: crates/editor/src/editor.rs (hunk 3 above)
Comment (wasd96040501) [reply crates/editor/src/editor.rs:7301]: I've just started working on this project and might not be very familiar with all aspects of it. How should I obtain the `ViewContext` from `AsyncWindowContext` for calling `r.range.to_offset`?

File: crates/editor/src/editor.rs (hunk 4 above)
Comment (wasd96040501) [reply crates/editor/src/editor.rs:7339]: Yes!

File: crates/editor/src/editor.rs (hunk 2 above)
Comment (wasd96040501) [reply crates/editor/src/editor.rs:7300]: Sure!

File: crates/editor/src/editor.rs (hunk 3 above)
Comment (osiewicz) [reply crates/editor/src/editor.rs:7301]: Hey, there's a `update` method on AsyncWindowContext that takes a `FnOnce(&mut WindowContext) -> R`.

**Review by SomeoneToIgnore — COMMENTED** [review]
Something's off with the way selections are influencing the lookup.
Consider anything as simple as 
```rust
struct Test;

fn test() {
    let test = Test;
}
```

1. I place the caret on the bottom `Test` and call the updated action — as expected, I navigate to the declaration, no extra window
2. The caret is moved to the `struct Test;` and selects the entire `Test` with the caret being at the end of this word
3. I call the action and expect it to go back, but instead it opens a separate tab, as the action did before the change
4. I go back and place the caret onto `struct Test;` myself, not touching the end of the word and call the action — everything works again

I would expect the action to never open a separate tab, even if it's in the end of the word, is that fair?
That's why it would be really beneficial to have a test for that: `editor_tests` contain examples of tests that set up fake language servers, we can reuse those to emulate corresponding LSP responses for our case and ensure there's no oddities like above.

**Comment by wasd96040501** [comment]
@SomeoneToIgnore 
Need help! It seems that "selects the entire `Test` with the caret" creates a new buffer. What can I do to figure out where this buffer from? So that I can compare where this buffer's from with target to fix this.

**Comment by SomeoneToIgnore** [comment]
I've added this on top of your PR
```diff
diff --git a/crates/editor/src/editor.rs b/crates/editor/src/editor.rs
index 6ca7bb6d2..719e1d69d 100644
--- a/crates/editor/src/editor.rs
+++ b/crates/editor/src/editor.rs
@@ -7311,18 +7311,21 @@ impl Editor {
             }
 
             // If there is one reference, just open it directly
-            if locations.len() == 1 {
+            if dbg!(locations.len()) == 1 {
                 let target = locations.pop().unwrap();
 
                 return editor.update(&mut cx, |editor, cx| {
                     let range = target.range.to_offset(target.buffer.read(cx));
                     let range = editor.range_for_match(&range);
 
-                    if Some(&target.buffer) == editor.buffer().read(cx).as_singleton().as_ref() {
+                    if Some(&target.buffer)
+                        == dbg!(editor.buffer().read(cx).as_singleton().as_ref())
+                    {
                         editor.change_selections(Some(Autoscroll::fit()), cx, |s| {
                             s.select_ranges([range]);
                         });
                     } else {
+                        dbg!("???????");
                         cx.window_context().defer(move |cx| {
                             let target_editor: View<Self> =
                                 workspace.update(cx, |workspace, cx| {
```

and did the same set of steps described in the review comment above, receiving this output:

```
[crates/editor/src/editor.rs:7314] locations.len() = 1
[crates/editor/src/editor.rs:7322] editor.buffer().read(cx).as_singleton().as_ref() = Some(
    Model { entity_id: EntityId(116v1), entity_type: "language::buffer::Buffer" },
)
[crates/editor/src/editor.rs:7314] locations.len() = 2
```

that last `[crates/editor/src/editor.rs:7314] locations.len() = 2` comes right before the things break and open the new multibuffer.
This seems to be right, looking at the surrounding code: we never hit the `if locations.len() == 1` condition added in this PR and go down to 
```rust
workspace.update(&mut cx, |workspace, cx| {
    let title = locations
        .first()
        .as_ref()
        .map(|location| {
            let buffer = location.buffer.read(cx);
            format!(
                "References to `{}`",
                buffer
                    .text_for_range(location.range.clone())
                    .collect::<String>()
            )
        })
        .unwrap();
    Self::open_locations_in_multibuffer(
        workspace, locations, replica_id, title, false, cx,
    );
})?;
```

code which does open the multibuffer.

Again, I would strongly advocate for writing a test for that, there's an `editor_tests.rs` file, consider `test_clone` as a donor for your new one:
https://github.com/zed-industries/zed/blob/939fa0d1ef7a37e51b63730251690bf326e2121a/crates/editor/src/editor_tests.rs#L472

It sets up a buffer with some contents and carets (`ˇ` symbol, there's multiple of them there, but we need only one), calls for actions:
```
editor.change_selections(None, cx, |s| s.select_ranges(selection_ranges.clone()));
```
and does some assertions.
We need to call our action and assert that the caret moves back and forth.

**Comment by normal-user-2** [comment]
Hey! Any update on this?

**Comment by wasd96040501** [comment]
> Hey! Any update on this?

Hah, will work on this these days.

File: crates/editor/src/editor.rs [hunk 5]
```diff
@@ -7287,17 +7287,68 @@ impl Editor {
...
+                        !r.range
+                            .to_offset(r.buffer.read(cx))
+                            .to_inclusive()
+                            .contains(&head_offset)
+                    })
+                    .map(|x| x.clone())
+                    .collect();
+            })?;
+
             if locations.is_empty() {
```
Comment (SomeoneToIgnore) [inline crates/editor/src/editor.rs:7320]: Current Zed opened multibuffer for such example, showing two items in it:
```rust
fn foo(param: i64) {
    let a = param;
    let a<|> = param;
    let a = param;
    println!(param);
}
```

I've expected it to toggle now, but now `locations.is_empty()` for such piece of code, this is a regression.
Let's add another test and fix it?

**Review by SomeoneToIgnore — COMMENTED** [review]
Nice, we're getting close to merging!

**Review by SomeoneToIgnore — COMMENTED** [review]
Thank you, I've tested it and it works like a charm for Rust.
One small gotcha we need to fix, here's the patch:
```diff
diff --git a/crates/editor/src/editor.rs b/crates/editor/src/editor.rs
index d7b0a629c..cc4875f6b 100644
--- a/crates/editor/src/editor.rs
+++ b/crates/editor/src/editor.rs
@@ -7287,7 +7287,6 @@ impl Editor {
         let buffer = self.buffer.read(cx);
         let head = self.selections.newest::<usize>(cx).head();
         let (buffer, head) = buffer.text_anchor_for_position(head, cx)?;
-        let head_offset = text::ToOffset::to_offset(&head, buffer.read(cx));
         let replica_id = self.replica_id(cx);
 
         let workspace = self.workspace()?;
@@ -7295,28 +7294,27 @@ impl Editor {
         let references = project.update(cx, |project, cx| project.references(&buffer, head, cx));
         Some(cx.spawn(|editor, mut cx| async move {
             let mut locations = references.await?;
+            let snapshot = buffer.update(&mut cx, |buffer, _| buffer.snapshot())?;
+            let head_offset = text::ToOffset::to_offset(&head, &snapshot);
 
-            // Lsp may return references that contain the item itself we requested `find_all_references` for (eg. rust-analyzer)
+            // LSP may return references that contain the item itself we requested `find_all_references` for (eg. rust-analyzer)
             // So we will remove it from locations
             // If there is only one reference, we will not do this filter cause it may make locations empty
             if locations.len() > 1 {
                 cx.update(|cx| {
-                    locations = locations
-                        .iter()
-                        .filter(|r| {
-                            // fn foo(x : i64) {
-                            //         ^
-                            //  println!(x);
-                            // }
-                            // It is ok to find reference when caret being at ^ (the end of the word)
-                            // So we turn offset into inclusive to include the end of the word
-                            !r.range
-                                .to_offset(r.buffer.read(cx))
-                                .to_inclusive()
-                                .contains(&head_offset)
-                        })
-                        .map(|x| x.clone())
-                        .collect();
+                    locations.retain(|location| {
+                        // fn foo(x : i64) {
+                        //         ^
+                        //  println!(x);
+                        // }
+                        // It is ok to find reference when caret being at ^ (the end of the word)
+                        // So we turn offset into inclusive to include the end of the word
+                        !location
+                            .range
+                            .to_offset(location.buffer.read(cx))
+                            .to_inclusive()
+                            .contains(&head_offset)
+                    });
                 })?;
             }
 
```

I've simplified the code a bit and, most imporantly, moved all snapshot retrieval after all .await points: otherwise, we risk transforming coordinates inforrectly, or even cause panics, e.g. https://github.com/zed-industries/zed/pull/8502

I was unable to push into your branch, and it's more than one line to apply, so would you mind to push it from your side?

Thank you for working on this.

**Comment by hamza72x** [comment]
CI failed, btw

https://github.com/zed-industries/zed/actions/runs/8093595834/job/22116562441

**Comment by SomeoneToIgnore** [comment]
Yes, old branch was not running some clippy checks, fixing.

**Comment by SomeoneToIgnore** [comment]
https://github.com/zed-industries/zed/pull/8590

**Comment by normal-user-2** [comment]
Is this reverted? If not: Then it's not working.

https://github.com/zed-industries/zed/assets/123471178/a87654d8-afac-4d1b-9ca8-07f1b6da5617

**Comment by normal-user-2** [comment]
```sh
Zed: v0.125.2 (Zed Preview)
OS: macOS 14.3.1
Memory: 16 GiB
Architecture: aarch64
```

**Comment by SomeoneToIgnore** [comment]
It's not reverted, but looks like there are more corner cases to consider.
Do you have a project with the example we can reproduce this issue with?

**Comment by SomeoneToIgnore** [comment]
Hmm, I actually do not see it in Preview despite working for me in Nightly: could be that it's not released yet, let's recheck tomorrow, after a new Preview is released.

**Comment by hamza-syndica** [comment]
https://github.com/normal-user-2/scrapper_rs

**Comment by maxbrunsfeld** [comment]
Sorry for the delayed response, but I think we need to revert this change. It's too inconsistent to sometimes open a multi-buffer and sometimes jump to a single location.

Also, omitting the current reference from the list, in the case where you open a multi-buffer is very misleading.

**Comment by hugh2slowmo** [comment]
@maxbrunsfeld 
> It's too inconsistent to sometimes open a multi-buffer and sometimes jump to a single location.

Most of the editors I have used, including VSCode, Vim, Emacs, and JetBrains, jump directly to the reference when there is only one reference. This greatly affects the user experience for those migrating over, making it feel very unintuitive.

When there is only one reference, 99% of the time after opening the multi-buffer, the operation is to move the cursor to that single reference and click to jump to the reference location in that file. This operation feels silly. What is the point of maintaining this interaction? Is it just for the sake of UI consistency?

**Comment by SomeoneToIgnore** [comment]
I also vote in favor of omitting the multi buffer.
So let's revive the PR but add a configuration knob to keep the old behavior with multi buffer open (disabled by default)

Happy to review that one.

**Comment by maxbrunsfeld** [comment]
If we re-do the change, we need to fix the implementation so that we don't eagerly filter out the current reference when we *do* show the multi-buffer.

**Pull request merged**